docs/contributing/error-handling.md

```diff
@@ -0,0 +1,366 @@
+
+To prevent this type of Terraform CLI error, the resource implementation should also check against `d.IsNewResource()` before removing from the Terraform State and returning no error. If that check is `true`, then remote operation error (or one synthesized from the non-existent status) should be returned instead. While adding this check will not fix the resource implementation to handle the eventually consistent nature of the remote system, the error being returned will be less opaque for operators and code maintainers to troubleshoot.
+
+In the Terraform AWS Provider, an initial fix for the Terraform CLI error will typically look like:
+
+```go
+func resourceServiceThingCreate(d *schema.ResourceData, meta interface{}) error {
+    /* ... */
+
+    return resourceServiceThingRead(d, meta)
+}
+
+func resourceServiceThingRead(d *schema.ResourceData, meta interface{}) error {
+    /* ... */
+
+    output, err := conn.DescribeServiceThing(input)
+
+    if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, "ResourceNotFoundException") {
+        log.Printf("[WARN] {Service} {Thing} (%s) not found, removing from state", d.Id())
+        d.SetId("")
+        return nil
+    }
+
+    if err != nil {
+        return fmt.Errorf("error reading {Service} {Thing} (%s): %w", d.Id(), err)
+    }
+
+    /* ... */
+}
+```
+
+Future documentation will show how to properly retry the remote operation for a short period of time until it is successful to remove the error completely.
+
+#### Creation Error Message Context
+
+Returning errors during creation should include additional messaging about the location or cause of the error for operators and code maintainers by wrapping with [`fmt.Errorf()`](https://pkg.go.dev/fmt#Errorf):
+
+```go
+if err != nil {
+    return fmt.Errorf("error creating {SERVICE} {THING}: %w", err)
+}
+```
+
+e.g.
+
+```go
+if err != nil {
+    return fmt.Errorf("error creating EC2 VPC: %w", err)
+}
+```
+
+Code that also uses waiters or other operations that return errors should follow a similar pattern, including the resource identifier since it has typically been set before this execution:
+
+```go
+if _, err := waiter.VpcAvailable(conn, d.Id()); err != nil {
+    return fmt.Errorf("error waiting for EC2 VPC (%s) availability: %w", d.Id(), err)
+}
+```
+
+### Resource Deletion
```

Comment:
content looks great! only thought is maybe move this Delete section after `Update` to keep with the usual logical order we have in resource schema where `Delete` is generally last

Reply:
Happy to change this if preferred, its currently in alphabetical order.

Reply:
ahh that makes sense! i'd think keep as is 👍